lib/states.js

```diff
@@ -29,50 +29,75 @@ Each "state" in the lifetime is a well defined function:
 
 @constructor
 @param {Array[Object]} hooks for handling states in the task lifecycle.
+@param {Stat} stats object, see stat.js
 */
 export default class States {
-  constructor(hooks) {
+  constructor(hooks, stats) {
     assert.ok(Array.isArray(hooks), 'hooks is an array');
     this.hooks = hooks;
+    this.stats = stats;
   }
 
   /**
   Invoke all hooks with a particular method.
   @param {Task} task handler.
   */
-  async _invoke(method, task) {
+  _invoke(method, task) {
     let hooks = this.hooks.filter(hasMethod.bind(this, method));
-    return  await Promise.all(hooks.map(async (hook) => {
-      return await hook[method](task);
-    }));
+    return this.stats.timeGen('tasks.time.states.' + method,
+      Promise.all(hooks.map(hook => hook[method](task)))
+    );
   }
 
   /**
-  The "link" state is responsible for creating any dependant containers and
-  returning the name of the container to be "linked" with the task container.
+  The "link" state is responsible for creating any dependent containers and
+  returning the name of the container to be "linked" with the task container,
+  additionally it may also return over-writeable environment variables to
+  give to the task.
 
-  Each "hook" which contains a link method _must_ return an array in the
-  following format:
+  Each "hook" which contains a link method or wants to declare an env variable
+  _must_ return an object in the following format:
 
   ```js
-      [
-        { name: 'container name', alias: 'alias in task container' }
+    {
+      links: [
+        {name: 'container name', alias: 'alias in task container'}
       ]
+      env: {
+        'name-of-env-var':  'value of variable'
+      },
+      binds: [
+        {source: '/path/on/host', target: '/path/in/container', readOnly: true}
+      ]
```

Comment:
I think collecting `links`, `env` and `binds` from features in the `link` hook makes sense...

We can now move "cache folders" and "encrypt env vars" into features... which would dramatically simplify the `task.js` file. And it would allow us to experiment with the rewrite/redesign of cache folders we've been talking about (preloaded read-only caches and all the other variants).